// quadstore/dist/cjs/serialization/quads.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.quadReader = exports.quadWriter = void 0;
const constants_1 = require("../utils/constants");
const terms_1 = require("./terms");
exports.quadWriter = {
    writtenValueLength: 0,
    write(prefix, value, baseValueOffset, quad, termNames, prefixes) {
        let ret = prefix;
        let valueOffset = baseValueOffset;
        for (let t = 0, term; t < termNames.length; t += 1) {
            term = quad[termNames[t]];
            ret += terms_1.termWriter.write(value, valueOffset, term, prefixes) + constants_1.separator;
            valueOffset += terms_1.termWriter.writtenValueLength;
        }
        this.writtenValueLength = valueOffset - baseValueOffset;
        return ret;
    },
};
exports.quadReader = {
    subject: null,
    predicate: null,
    object: null,
    graph: null,
    read(key, keyOffset, value, valueOffset, termNames, factory, prefixes) {
        for (let t = 0, termName; t < termNames.length; t += 1) {
            termName = termNames[t];
            this[termName] = terms_1.termReader.read(key, keyOffset, value, valueOffset, factory, prefixes);
            keyOffset += terms_1.termReader.readKeyChars + constants_1.separator.length;
            valueOffset += terms_1.termReader.readValueLength;
        }
        return factory.quad(this.subject, this.predicate, this.object, this.graph);
    },
};
